traffic-advice.bs

```diff
@@ -0,0 +1,265 @@
+
+An <dfn>agent identity</dfn> is a [=list=] of [=strings=]. It must contain at least two elements, and the last must be `"*"`.
+
+Identity {#identity}
+====================
+
+Each agent should have an brand name that specifically identifies it (such as `PollyPrefetchProxy`).
+
+Its [=agent identity=] is all of the following that apply, in order:
+
+1.  The brand name
+1.  `"prefetch-proxy"`, if the agent is a proxy server which exclusively serves prefetch traffic (for example, a [private prefetch proxy](https://github.com/buettner/private-prefetch-proxy))
+1.  `"*"`
+
+Fetching {#fetching}
+====================
+
+<section algorithm="generate-url">
+
+To <dfn>generate a traffic advice URL</dfn> for [=origin=] |origin|, run the following steps:
+
+1.  If |origin| is not a [=tuple origin=], return failure.
+
+1.  If |origin|'s [=origin/scheme=] is not an [=HTTP(S) scheme=], return failure.
+
+1. If |origin| is not a [=potentially trustworthy origin=], return failure.
+
+1.  Return a new [=URL=] as follows:
+
+    :  [=url/scheme=]
+    :: |origin|'s [=origin/scheme=]
+    :  [=url/host=]
+    :: |origin|'s [=origin/host=]
+    :  [=url/port=]
+    :: |origin|'s [=origin/port=]
+    :  [=url/path=]
+    :: « `".well-known"`, `"traffic-advice"` »
+    
+</section>
+
+<section algorithm="fetch-traffic-advice">
+
+To <dfn>fetch traffic advice</dfn> for [=origin=] |origin|, [=agent identity=] |identity| and algorithm |whenComplete| accepting a [=traffic advice result=]:
+
+1.  Let |url| be the result of [=generating a traffic advice URL=] for |origin|.
+    If it results in failure, then return failure.
+
+1.  Let |request| be a [=request=] as follows:
+
+    :  [=request/method=]
+    :: `` `GET` ``
+    :  [=request/URL=]
+    :: |url|
+    :  [=request/client=]
+    :: null
+    :  [=request/credentials mode=]
+    :: `"omit"`
+    :  [=request/redirect mode=]
+    :: `"manual"`
+        <div class="note">This means that a [=redirect status=] will not lead to another origin being contacted.</div>
+
+1.  Let |processResponse| be the following steps, given [=response=] |response|:
+
+    1.  If |response|'s [=response/type=] is `"error"`, then terminate the fetch, run |whenComplete| with `"unreachable"`, and return.
+
+    1.  If |response|'s [=response/type=] is `"opaqueredirect"`, then terminate the fetch, run |whenComplete| with null, and return.
+
+    1.  [=Assert=]: |response|'s [=response/type=] is `"basic"`.
+
+    1.  If |response|'s [=response/status=] is 429 (Too Many Requests; see [[RFC6585]]) or 503 (Service Unavailable; see [[HTTP-SEMANTICS]]), then terminate the fetch, run |whenComplete| with `"unreachable"`, and return.
+        <div class="note">If present, the [[HTTP-SEMANTICS]] `Retry-After` response header could be used as a hint about when to next retry.</div>
+
+    1.  If |response|'s [=response/status=] is not an [=ok status=], then terminate the fetch, run |whenComplete| with null and return.
```

Comment:
/cc @annevk to note that with the new-ish fetch architecture processResponse callbacks need to terminate the fetch manually for non-network error errors (especially to avoid later calling processResponseDone). That's kind of annoying, although I'm not sure I can think of anything better.

Reply:
We could make "require ok" part of fetch, but that might not always do the right thing for resource timing. Or at least we'd have to better understand that first.

Reply:
Yeah, you want something more generic to handle e.g. MIME type mismatches.

Hmm, maybe process response could return a value, and if it returns false then fetch does not call processResponseDone? Although, I'm not sure "return false" is any better than "terminate the fetch".